moveit_core/robot_state/test/attached_body_test.cpp

```diff
@@ -0,0 +1,133 @@
+#include <moveit/robot_model/robot_model.hpp>
+#include <moveit/robot_state/robot_state.hpp>
+#include <urdf_parser/urdf_parser.h>
+#include <gtest/gtest.h>
+
+class SingleLinkRobot : public ::testing::Test
+{
+public:
+  SingleLinkRobot()
+  {
+    static const std::string URDF_XML = R"(
+      <?xml version="1.0" ?>
+      <robot name="single_link_robot">
+      <link name="link">
+        <inertial>
+          <mass value="2.81"/>
+          <origin rpy="0 0 0" xyz="0.0 0.0 .0"/>
+          <inertia ixx="0.1" ixy="-0.2" ixz="0.5" iyy="-.09" iyz="1" izz="0.101"/>
+        </inertial>
+        <collision name="my_collision">
+          <origin rpy="0 0 0" xyz="0 0 0"/>
+          <geometry>
+            <box size="1 2 1" />
+          </geometry>
+        </collision>
+        <visual>
+          <origin rpy="0 0 0" xyz="0.0 0 0"/>
+          <geometry>
+            <box size="1 2 1" />
+          </geometry>
+        </visual>
+      </link>
+      </robot>
+    )";
+
+    static const std::string SRDF_XML = R"xml(
+      <?xml version="1.0" ?>
+      <robot name="single_link_robot">
+      </robot>
+      )xml";
+
+    urdf::ModelInterfaceSharedPtr urdf_model = urdf::parseURDF(URDF_XML);
+    auto srdf_model = std::make_shared<srdf::Model>();
+    srdf_model->initString(*urdf_model, SRDF_XML);
+    robot_model_ = std::make_shared<moveit::core::RobotModel>(urdf_model, srdf_model);
+  }
+
+protected:
+  moveit::core::RobotModelConstPtr robot_model_;
+};
+
+class SingleAttachedBody : public SingleLinkRobot
+{
+public:
+  SingleAttachedBody()
+  {
+    const moveit::core::LinkModel* link = robot_model_->getLinkModel("link");
+    std::string name = "root_body";
+    Eigen::Isometry3d pose;
+    pose = Eigen::Translation3d(1, 2, 3);
+    auto box = std::make_shared<shapes::Box>(0.1, 0.2, 0.3);
+    std::vector<shapes::ShapeConstPtr> shapes = { box };
+    EigenSTL::vector_Isometry3d shape_poses;
+    Eigen::Isometry3d shape_pose;
+    shape_pose = Eigen::Translation3d(4, 5, 6);
+    shape_poses.push_back(shape_pose);
+    std::set<std::string> touch_links = { "link" };
+    trajectory_msgs::msg::JointTrajectory detach_posture;
+    detach_posture.joint_names.push_back("joint");
+    trajectory_msgs::msg::JointTrajectoryPoint p;
+    p.positions.push_back(0.1);
+    detach_posture.points.push_back(p);
+    Eigen::Isometry3d subframe_pose;
+    subframe_pose = Eigen::Translation3d(7, 8, 9);
+    moveit::core::FixedTransformsMap subframes{ { "subframe", subframe_pose } };
+    root_body_ = std::make_shared<moveit::core::AttachedBody>(link, name, pose, shapes, shape_poses, touch_links,
+                                                              detach_posture, subframes);
+  }
+
+protected:
+  std::shared_ptr<moveit::core::AttachedBody> root_body_;
+};
+
+// Verifies that a single body attached to a link works as intended
+
+TEST_F(SingleAttachedBody, RootBodyHasCorrecAttachedLink)
+{
+  ASSERT_EQ(root_body_->getName(), "root_body");
```

Comment:
Should this test try to actually attach `root_body_` to the robot model, and then do some lookups **on** the robot model to verify that the information propagated correctly?

Looking at the code base, seems like we want to
* Get a robot state from the robot model
* Call `RobotState::attachBody`
* Call `RobotState::getAttachedBodies` and grab the info
* Clear the bodies with `RobotState::clearAttachedBody`
* Verify again the object is gone, etc.